Delete setters

jijiew opened the pull request:
This PR implement the delete setters with the following cmd

delete name

It mainly delete the setter from open api file and remove all comment from the resource files.

if the setter to be deleted is used in substitution, it will return error

jijiew commented:
/cc mortent

jijiew commented:
/cc phanimarupaka

jijiew commented:
/assign @liujingfang1

mortent commented:
So I think the code here looks good. But this does remove the setter completely from the Kptfile, so what happens when someone deletes a setter from their own local version of a package, and then updates it with a new version from upstream? I think the desired behavior would be that the setter is not added back in this case, so I think we might need a tombstone after deletion.

mortent commented:
/ok-to-test

phanimarupaka commented:
@mortent As we do 3-way merge, tombstone is not required. If a setter is deleted then it is identified as deleted when compared with original and will not add it back.

@jijiew You may add openAPI file test cases here https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/merge3/merge3_test.go#L14,  to confirm the behavior.

original - should have an example setter1
updated - same as original
local - delete setter1
expected- make sure setter1 is not added back

mortent commented:
@phanimarupaka I have tested this and the setter is added back. Actually seems like it gets added back in the Kptfile but not the in the resource manifest, so we end up in a weird state.

jijiew commented:
@mortent @phanimarupaka I also tested in kpt side, it does get added back. For the resource file in kustomized side using the merge3, from the test there it seams Okay to me (as delete only remove the comment).
as @mortent said, it seams to get add back here:
https://github.com/GoogleContainerTools/kpt/blob/master/internal/kptfile/pkgfile.go
where it merges the kpt file. 
in that case, we could probably keep this pr and add change to kpt side after release. But I will double check

jijiew commented:
@mortent @phanimarupaka 
I have a separate PR on Kpt side solving the kptfile merge issue:
https://github.com/GoogleContainerTools/kpt/pull/679

phanimarupaka commented:
@mortent @jijiew You are right. The section I pointed to was for resource files which is already handled. But Kptfile does it a bit different. So the fix should be on Kpt repo to do the merge to not add the setters back.

mortent commented on kyaml/setters2/delete_test.go:
```diff
@@ -0,0 +1,192 @@
+ `,
+		},
+		{
+			name:        "set-foo-type",
```
Are the names and descriptions here correct? They don't look like they have anything to do with deletion.

jijiew replied:
Done.

mortent commented on cmd/config/internal/commands/cmddelete_test.go:
```diff
@@ -0,0 +1,147 @@
+		err               string
+	}{
+		{
+			name: "delete replicas",
```
Maybe we could add some additional cases here? Like deleting a setter when there are other setters in the openAPI file and deletion of a setter that doesn't exist.

jijiew replied:
Ack. & Done.

phanimarupaka commented on cmd/config/internal/commands/cmddelete_test.go:
```diff
@@ -0,0 +1,244 @@
+metadata:
+  name: nginx-deployment
+spec:
+  replicas: 3 # {"$ref":"#/definitions/io.k8s.cli.setters.replicas"}
```
Please add test case for shorthand ref as well. {"$openapi" : "replicas"}. Refer to cmdcreatesetter_test.go

jijiew replied:
I changed it to use replicas since that's the new format

phanimarupaka commented on kyaml/fieldmeta/fieldmeta.go:
```diff
@@ -151,10 +151,15 @@ func (fm *FieldMeta) Write(n *yaml.RNode) error {
-	// {"openAPI":"replicas"} and added to the line comment
-	arr := strings.Split(fm.Schema.Ref.String(), ".")
-	n.YNode().LineComment = fmt.Sprintf(`{"%s":"%s"}`, shortHandRef, arr[len(arr)-1])
+	if fm.Schema.Ref.String() != "" {
```
Can you explain the idea behind this in a comment.

jijiew replied:
Done.

phanimarupaka commented:
@jijiew Does this also delete substitutions ? I think we should delete substitutions as well as part of this command. However, we need to think about the UX here. Does `delete-setter` seem to be a write name for it ?

`create-setter` and `create-subst` are different commands currently as they are created differently. But delete can just take in the name and delete either setter or substitution as they can't share same name. If we decide to delete substitution, then we should think about, how should we be deleting underlying setters related to that substitution(it gets complicated in the case of [nested substitutions](https://github.com/kubernetes-sigs/kustomize/pull/2551)). We should also think about the behavior of orphaned references/definitions as mentioned in this issue. https://github.com/GoogleContainerTools/kpt/issues/674

Tagging @frankfarzan for UX decision.

jijiew commented:
> Jijie Wei Does this also delete substitutions ? I think we should delete substitutions as well as part of this command. However, we need to think about the UX here. Does `delete-setter` seem to be a write name for it ?
> 
> `create-setter` and `create-subst` are different commands currently as they are created differently. But delete can just take in the name and delete either setter or substitution as they can't share same name. If we decide to delete substitution, then we should think about, how should we be deleting underlying setters related to that substitution(it gets complicated in the case of [nested substitutions](https://github.com/kubernetes-sigs/kustomize/pull/2551)). We should also think about the behavior of orphaned references/definitions as mentioned in this issue. [GoogleContainerTools/kpt#674](https://github.com/GoogleContainerTools/kpt/issues/674)
> 
> Tagging Frank Farzan for UX decision.

Agree, this pr does not handle deletion in substitution as the is put in the description. It looks to me that delete a setter would implicitly delete substituition as well. what you guys think?

frankfarzan commented:
Can you create an issue and discuss the proposed UX and its implications? It should be in a form of user guide that should becomes the doc changes for this functionality.

jijiew commented:
> Can you create an issue and discuss the proposed UX and its implications? It should be in a form of user guide that should becomes the doc changes for this functionality.
@phanimarupaka @mortent 
created a new issue #2569 please take a look
also updated the pr:
1. to use delete-setter
2. updated the current pr so that when the setter to be deleted is used in substitution, it will return error

jijiew commented on cmd/config/internal/commands/cmddeletesetter.go:
```diff
@@ -0,0 +1,64 @@
+	c := &cobra.Command{
+		Use:     "delete-setter DIR NAME",
+		Args:    cobra.MinimumNArgs(2),
+		Short:   commands.SetShort,
```
Done.

phanimarupaka commented on kyaml/setters2/delete.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, errors.Errorf("setter does not exist")
+	}
+
+	subst := ContainedInSubstituion(definitions, key)
```
Better to rename method to something like `SubstReferringSetter` as the current name looks like it returns boolean.

jijiew replied:
Done.

phanimarupaka commented on cmd/config/internal/commands/cmddeletesetter_test.go:
```diff
@@ -0,0 +1,304 @@
+apiVersion: v1alpha1
+kind: Example
+openAPI:
+  definitions: {}
```
If there are no definitions then it's better to remove this section completely. This would create json definitions upon creating a new setter. Try creating a setter with this file and you can see the difference.

jijiew replied:
I see, removed, also removed the empty openApi

phanimarupaka commented on kyaml/setters2/delete.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, errors.Errorf("setter is used in substitution %s, please delete the substitution first", subst)
+	}
+
+	for i := 0; i < len(definitions.Content()); i += 2 {
```
Is this section takes a setter key and deletes it from openAPI definitions? In that case, can't we use FieldClearer here and make it as simple as https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/kio/byteio_writer.go#L75

jijiew replied:
good point, thanks

phanimarupaka commented on kyaml/setters2/delete.go:
```diff
@@ -0,0 +1,152 @@
+				continue
+			}
+
+			b, err := substNode.String()
```
Nit: b usually refers to bytes, but it is string here.

jijiew replied:
Done.

phanimarupaka commented:
@jijiew This looks good. Thanks for the PR.

phanimarupaka commented:
/lgtm
/approve
